loom/engine/bindings/loom/lmSQLite.cpp

```diff
@@ -0,0 +1,467 @@
+//                    into the onStatementComplete delegate... which i'm still not sure why it is
+//                    done there but not for onStatementProgress...
+//
+//HACK: NOT STABLE!!!!
+stepAsyncLuaState = L;
+        loom_thread_start(Statement::stepAsyncBody, this);    
+    }
+
+
+    const char *columnName(int col)
+    {
+        return sqlite3_column_name(statementHandle, col);
+    }
+
+    int columnType(int col)
+    {
+        return sqlite3_column_type(statementHandle, col);
+    }
+
+    int columnInt(int col)
+    {
+        return sqlite3_column_int(statementHandle, col);
+    }
+
+    double columnDouble(int col)
+    {
+        return sqlite3_column_double(statementHandle, col);
+    }
+
+    const char* columnString(int col)
+    {
+        return (const char *)sqlite3_column_text(statementHandle, col);
+    }
+
+    utByteArray *columnBytes(int col)
+    {
+        //get the blob from the column
+        void *blob = (void *)sqlite3_column_blob(statementHandle, col);
+        if(blob == NULL)
+        {
+            return NULL;
+        }
+        int size = sqlite3_column_bytes(statementHandle, col);
+
+        //valid blob so allocate byte array for it
+        utByteArray *bytes = new utByteArray();
+        bytes->allocateAndCopy(blob, size);     
+
+        return bytes;   
+    }
+
+    int reset()
+    {
+        int result = sqlite3_reset(statementHandle); 
+        if(result != SQLITE_OK)
+        {
+            lmLogError(gSQLiteGroup, "Error calling reset for database: %s with Result Code: %i", parentDB->getDBName(), result);
+        }
+        return result;        
+    }
+
+    int finalize()
+    {
+        int result = sqlite3_finalize(statementHandle); 
+        if(result != SQLITE_OK)
+        {
+            lmLogError(gSQLiteGroup, "Error calling finalize for database: %s with Result Code: %i", parentDB->getDBName(), result);
+        }
+        return result;        
+    }
+};
+
+
+//---Statement--- external function initialisation
+int __stdcall Statement::stepAsyncBody(void *param)
+{
+    //get the statement
+    Statement *s = (Statement *)param;
+
+
+    //call the internal SQLite step function, then yield the thread
+    int result = sqlite3_step(s->statementHandle); 
+    loom_thread_yield();
+
+    //mutex so we can call the delegate
+    s->stepAsyncDelegateMutex = loom_mutex_create();
+    loom_mutex_lock(s->stepAsyncDelegateMutex);
+
+    //check the result now
+    if(result == SQLITE_ROW)
+    {
+        s->_OnStatementProgressDelegate.invoke();
+    }
+    else if(result == SQLITE_DONE)
+    {
+//HACK: s->stepAsyncLuaState is not working as it can be unstable... 
+//      need another way to pass the Statement through to OnStatementComplete!!!
+lualoom_pushnative<Statement>(s->stepAsyncLuaState, s);
+s->_OnStatementCompleteDelegate.incArgCount();
+s->_OnStatementCompleteDelegate.invoke();
+lua_pop(s->stepAsyncLuaState, 1);
```

Comment:
It's cool to keep the param if the L solution works.

Should onProgress not have the Statement as a param also though?

Sorry, SQLite is strange and frightening...
On 12 Feb 2015 10:58 PM, "Ben Garney" <notifications@github.com> wrote:

> In loom/engine/bindings/loom/lmSQLite.cpp
> <https://github.com/LoomSDK/LoomSDK/pull/261#discussion_r24619496>:
>
> > +    s->stepAsyncDelegateMutex = loom_mutex_create();
> > +    loom_mutex_lock(s->stepAsyncDelegateMutex);
> > +
> > +    //check the result now
> > +    if(result == SQLITE_ROW)
> > +    {
> > +        s->_OnStatementProgressDelegate.invoke();
> > +    }
> > +    else if(result == SQLITE_DONE)
> > +    {
> > +//HACK: s->stepAsyncLuaState is not working as it can be unstable...
> > +//      need another way to pass the Statement through to OnStatementComplete!!!
> > +lualoom_pushnative<Statement>(s->stepAsyncLuaState, s);
> > +s->_OnStatementCompleteDelegate.incArgCount();
> > +s->_OnStatementCompleteDelegate.invoke();
> > +lua_pop(s->stepAsyncLuaState, 1);
>
> Is this needed? Ahh yeah cuz you don't have a note.
>
> One option here is that we could have an "inner" progress delegate that
> goes through LS to reassociate the statement and pass it as a param?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LoomSDK/LoomSDK/pull/261/files#r24619496>.
>